src/execution_plan/ops/op_create.c

```diff
@@ -104,41 +105,89 @@ void _CreateEdges(OpCreate *op, Record r) {
 /* Commit insertions. */
 static void _CommitNodes(OpCreate *op) {
 	Node *n;
-	int labelID;
-	Graph *g = op->gc->g;
+	GraphContext *gc = op->gc;
+	Graph *g = gc->g;
+
+	/* Create missing schemas.
+	 * this loop iterates over the CREATE pattern, e.g.
+	 * CREATE (p:Person)
+	 * As such we're not expecting a large number of iterations. */
+	uint blueprint_node_count = array_len(op->nodes_to_create);
+	for(uint i = 0; i < blueprint_node_count; i++) {
+		NodeCreateCtx *node_ctx = op->nodes_to_create + i;
+
+		const char *label = node_ctx->node->label;
+		if(label) {
+			if(GraphContext_GetSchema(gc, label, SCHEMA_NODE) == NULL) {
+				Schema *s = GraphContext_AddSchema(gc, label, SCHEMA_NODE);
+				op->stats->labels_added++;
+			}
+		}
+
+		// Introduce attributes to graph.
+		if(!node_ctx->properties) continue;
+		uint prop_count = node_ctx->properties->property_count;
+		for(int prop_idx = 0; prop_idx < prop_count; prop_idx ++) {
+			const char *prop = node_ctx->properties->keys[prop_idx];
+			GraphContext_FindOrAddAttribute(gc, prop);
+		}
+	}
```

Comment:
This logic (and the corresponding edge logic) seems complex enough to warrant a separate call. Ideally in a place where OpMerge can access it too, since there should be no difference in the work the two calls are doing if Merge is in a create context. (Though that doesn't necessarily have to be in this PR, I have clearly put it off too!)